src/core/lib/iomgr/work_serializer.cc

```diff
@@ -93,39 +128,60 @@ void WorkSerializer::WorkSerializerImpl::Orphan() {
 }
 
 // The thread that calls this loans itself to the work serializer so as to
-// execute all the scheduled callback. This is called from within
-// WorkSerializer::Run() after executing a callback immediately, and hence size_
-// is at least 1.
+// execute all the scheduled callbacks.
 void WorkSerializer::WorkSerializerImpl::DrainQueue() {
+  if (GRPC_TRACE_FLAG_ENABLED(grpc_work_serializer_trace)) {
+    gpr_log(GPR_INFO, "WorkSerializer::DrainQueue() %p", this);
+  }
+  // Attempt to take ownership of the WorkSerializer.
+  const uint64_t prev_ref_pair =
+      refs_.fetch_add(MakeRefPair(1, 1), std::memory_order_acq_rel);
+  if (GetOwners(prev_ref_pair) == 0) {
+    // We took ownership of the WorkSerializer. Drain the queue.
+    DrainQueueOwned();
+  } else {
+    // Another thread is holding the WorkSerializer, so decrement the ownership
+    // count we just added and queue a no-op callback.
+    refs_.fetch_sub(MakeRefPair(1, 0), std::memory_order_acq_rel);
+    CallbackWrapper* cb_wrapper = new CallbackWrapper([]() {}, DEBUG_LOCATION);
```

Comment:
Can we avoid an allocation here by using a sentinel value?  For example, we could add an `MultiProducerSingleConsumerQueue::Node sentinel_node` data member to `WorkSerializerImpl` and push that onto the queue, and then have `DrainQueueOwned()` check for that value and ignore it.

Reply:
If we used a single copy of sentinel node, it would cause issues if multiple threads invoke `DrainQueue()` when a thread already owns a lock to the WorkSerializer messing up the internals of the mpscq.

Reply:
Good catch.  That's unfortunate, but I suspect it's acceptable in the cases where we want to use `DrainQueue()`.